BackfillEthereumTransactionHashMigration failed due to db role permission

### Description

BackfillEthereumTransactionHashMigration failed in staging:

```
2024-08-20T14:33:15.729Z ERROR main o.s.b.SpringApplication Application run failed org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaConfiguration.class]: Failed to initialize dependency 'flywayInitializer' of LoadTimeWeaverAware bean 'entityManagerFactory': Error creating bean with name 'flywayInitializer' defined in class path resource [org/springframework/boot/autoconfigure/flyway/FlywayAutoConfiguration$FlywayConfiguration.class]: Migration failed !
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:326)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:205)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:954)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:625)
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:66)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:754)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:456)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:335)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1363)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1352)
	at com.hedera.mirror.importer.ImporterApplication.main(ImporterApplication.java:28)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:91)
	at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:53)
	at org.springframework.boot.loader.launch.JarLauncher.main(JarLauncher.java:58)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'flywayInitializer' defined in class path resource [org/springframework/boot/autoconfigure/flyway/FlywayAutoConfiguration$FlywayConfiguration.class]: Migration failed !
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1806)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:600)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:313)
	... 15 common frames omitted
Caused by: org.flywaydb.core.internal.command.DbMigrate$FlywayMigrateException: Migration failed !
	at org.flywaydb.core.internal.command.DbMigrate.doMigrateGroup(DbMigrate.java:382)
	at org.flywaydb.core.internal.command.DbMigrate.lambda$applyMigrations$1(DbMigrate.java:272)
	at org.flywaydb.core.internal.jdbc.TransactionalExecutionTemplate.execute(TransactionalExecutionTemplate.java:55)
	at org.flywaydb.core.internal.command.DbMigrate.applyMigrations(DbMigrate.java:271)
	at org.flywaydb.core.internal.command.DbMigrate.migrateGroup(DbMigrate.java:244)
	at org.flywaydb.core.internal.command.DbMigrate.lambda$migrateAll$0(DbMigrate.java:139)
	at org.flywaydb.database.postgresql.PostgreSQLAdvisoryLockTemplate.execute(PostgreSQLAdvisoryLockTemplate.java:73)
	at org.flywaydb.database.postgresql.PostgreSQLAdvisoryLockTemplate.lambda$execute$0(PostgreSQLAdvisoryLockTemplate.java:56)
	at org.flywaydb.core.internal.jdbc.TransactionalExecutionTemplate.execute(TransactionalExecutionTemplate.java:55)
	at org.flywaydb.database.postgresql.PostgreSQLAdvisoryLockTemplate.execute(PostgreSQLAdvisoryLockTemplate.java:56)
	at org.flywaydb.database.postgresql.PostgreSQLConnection.lock(PostgreSQLConnection.java:96)
	at org.flywaydb.core.internal.schemahistory.JdbcTableSchemaHistory.lock(JdbcTableSchemaHistory.java:145)
	at org.flywaydb.core.internal.command.DbMigrate.migrateAll(DbMigrate.java:139)
	at org.flywaydb.core.internal.command.DbMigrate.migrate(DbMigrate.java:97)
	at org.flywaydb.core.Flyway.lambda$migrate$0(Flyway.java:202)
	at org.flywaydb.core.FlywayExecutor.execute(FlywayExecutor.java:205)
	at org.flywaydb.core.Flyway.migrate(Flyway.java:147)
	at org.springframework.boot.autoconfigure.flyway.FlywayMigrationInitializer.afterPropertiesSet(FlywayMigrationInitializer.java:66)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1853)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1802)
	... 22 common frames omitted
Caused by: org.flywaydb.core.api.FlywayException: Migration failed !
	at org.flywaydb.core.internal.resolver.java.JavaMigrationExecutor.executeOnce(JavaMigrationExecutor.java:76)
	at org.flywaydb.core.internal.resolver.java.JavaMigrationExecutor.lambda$execute$0(JavaMigrationExecutor.java:54)
	at org.flywaydb.core.internal.database.DefaultExecutionStrategy.execute(DefaultExecutionStrategy.java:27)
	at org.flywaydb.core.internal.resolver.java.JavaMigrationExecutor.execute(JavaMigrationExecutor.java:53)
	at org.flywaydb.core.internal.command.DbMigrate.doMigrateGroup(DbMigrate.java:374)
	... 41 common frames omitted
Caused by: org.springframework.jdbc.BadSqlGrammarException: PreparedStatementCallback; bad SQL grammar [with deleted as (
  delete from contract_transaction_hash
  where consensus_timestamp = ? and hash = ''::bytea
  returning *
)
insert into contract_transaction_hash (consensus_timestamp, entity_id, hash, payer_account_id, transaction_result)
select consensus_timestamp, entity_id, ?, payer_account_id, transaction_result
from deleted
]
	at org.springframework.jdbc.support.SQLStateSQLExceptionTranslator.doTranslate(SQLStateSQLExceptionTranslator.java:112)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:107)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:116)
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1548)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:677)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:701)
	at org.springframework.jdbc.core.JdbcTemplate.batchUpdate(JdbcTemplate.java:1101)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:354)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:716)
	at org.springframework.jdbc.core.JdbcTemplate$$SpringCGLIB$$0.batchUpdate(<generated>)
	at com.hedera.mirror.importer.migration.BackfillEthereumTransactionHashMigration.backfillTables(BackfillEthereumTransactionHashMigration.java:157)
	at com.hedera.mirror.importer.migration.BackfillEthereumTransactionHashMigration.lambda$doMigrate$3(BackfillEthereumTransactionHashMigration.java:137)
	at org.springframework.transaction.support.TransactionOperations.lambda$executeWithoutResult$0(TransactionOperations.java:68)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140)
	at org.springframework.transaction.support.TransactionOperations.executeWithoutResult(TransactionOperations.java:67)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:354)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:216)
	at jdk.proxy2/jdk.proxy2.$Proxy165.executeWithoutResult(Unknown Source)
	at com.hedera.mirror.importer.migration.BackfillEthereumTransactionHashMigration.doMigrate(BackfillEthereumTransactionHashMigration.java:117)
	at com.hedera.mirror.importer.migration.AbstractJavaMigration.migrate(AbstractJavaMigration.java:54)
	at com.hedera.mirror.importer.migration.BackfillEthereumTransactionHashMigration.migrate(BackfillEthereumTransactionHashMigration.java:42)
	at org.flywaydb.core.internal.resolver.java.JavaMigrationExecutor.executeOnce(JavaMigrationExecutor.java:62)
	... 45 common frames omitted
Caused by: java.sql.BatchUpdateException: Batch entry 0 with deleted as (
  delete from contract_transaction_hash
  where consensus_timestamp = ('1722306953589947003'::int8) and hash = ''::bytea
  returning *
)
insert into contract_transaction_hash (consensus_timestamp, entity_id, hash, payer_account_id, transaction_result)
select consensus_timestamp, entity_id, ?, payer_account_id, transaction_result
from deleted
 was aborted: ERROR: permission denied for table contract_transaction_hash  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:165)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2413)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:579)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:912)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:936)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1733)
	at com.zaxxer.hikari.pool.ProxyStatement.executeBatch(ProxyStatement.java:127)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeBatch(HikariProxyPreparedStatement.java)
	at org.springframework.jdbc.core.JdbcTemplate.lambda$batchUpdate$4(JdbcTemplate.java:1117)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:658)
	... 66 common frames omitted
Caused by: org.postgresql.util.PSQLException: ERROR: permission denied for table contract_transaction_hash
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2725)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2412)
	... 74 common frames omitted
```

### Steps to reproduce

check the description

### Additional context

_No response_

### Hedera network

other

### Version

v0.112.0-rc1

### Operating system

None

--- a/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/BackfillEthereumTransactionHashMigration.java
+++ b/hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/BackfillEthereumTransactionHashMigration.java
@@ -22,12 +22,13 @@ import com.google.common.base.Stopwatch;
 import com.hedera.mirror.common.domain.entity.EntityId;
 import com.hedera.mirror.common.domain.transaction.TransactionHash;
 import com.hedera.mirror.importer.ImporterProperties;
+import com.hedera.mirror.importer.config.Owner;
 import com.hedera.mirror.importer.parser.record.entity.EntityProperties;
 import com.hedera.mirror.importer.parser.record.ethereum.EthereumTransactionParser;
-import com.hedera.mirror.importer.repository.TransactionHashRepository;
 import jakarta.inject.Named;
 import java.io.IOException;
 import java.util.List;
+import java.util.Objects;
 import java.util.concurrent.atomic.AtomicInteger;
 import lombok.Data;
 import org.apache.commons.lang3.ArrayUtils;
@@ -37,11 +38,18 @@ import org.springframework.jdbc.core.DataClassRowMapper;
 import org.springframework.jdbc.core.JdbcTemplate;
 import org.springframework.jdbc.core.ParameterizedPreparedStatementSetter;
 import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.datasource.DataSourceTransactionManager;
 import org.springframework.transaction.support.TransactionOperations;
+import org.springframework.transaction.support.TransactionTemplate;
 
 @Named
 public class BackfillEthereumTransactionHashMigration extends RepeatableMigration {
 
+    private static final String INSERT_TRANSACTION_HASH_SQL =
+            """
+            insert into transaction_hash (consensus_timestamp, distribution_id, hash, payer_account_id)
+            values (?, ?, ?, ?)
+            """;
     private static final ParameterizedPreparedStatementSetter<MigrationEthereumTransaction> PSS = (ps, transaction) -> {
         ps.setBytes(1, transaction.getHash());
         ps.setLong(2, transaction.getConsensusTimestamp());
@@ -84,23 +92,17 @@ public class BackfillEthereumTransactionHashMigration extends RepeatableMigratio
     private final EntityProperties entityProperties;
     private final EthereumTransactionParser ethereumTransactionParser;
     private final JdbcTemplate jdbcTemplate;
-    private final TransactionHashRepository transactionHashRepository;
-    private final TransactionOperations transactionOperations;
 
     @Lazy
     public BackfillEthereumTransactionHashMigration(
             EntityProperties entityProperties,
             EthereumTransactionParser ethereumTransactionParser,
             ImporterProperties importerProperties,
-            JdbcTemplate jdbcTemplate,
-            TransactionHashRepository transactionHashRepository,
-            TransactionOperations transactionOperations) {
+            @Owner JdbcTemplate jdbcTemplate) {
         super(importerProperties.getMigration());
         this.entityProperties = entityProperties;
         this.ethereumTransactionParser = ethereumTransactionParser;
         this.jdbcTemplate = jdbcTemplate;
-        this.transactionHashRepository = transactionHashRepository;
-        this.transactionOperations = transactionOperations;
     }
 
     @Override
@@ -114,7 +116,7 @@ public class BackfillEthereumTransactionHashMigration extends RepeatableMigratio
         var patched = new AtomicInteger();
         var stopwatch = Stopwatch.createStarted();
 
-        transactionOperations.executeWithoutResult(s -> {
+        getTransactionOperations().executeWithoutResult(s -> {
             long consensusTimestamp = -1;
             for (; ; ) {
                 var transactions = jdbcTemplate.query(SELECT_ETHEREUM_TRANSACTION_SQL, ROW_MAPPER, consensusTimestamp);
@@ -173,8 +175,19 @@ public class BackfillEthereumTransactionHashMigration extends RepeatableMigratio
                             .payerAccountId(t.getPayerAccountId())
                             .build())
                     .toList();
-            transactionHashRepository.saveAll(transactionHashes);
+            jdbcTemplate.batchUpdate(
+                    INSERT_TRANSACTION_HASH_SQL, transactionHashes, transactionHashes.size(), (ps, transactionHash) -> {
+                        ps.setLong(1, transactionHash.getConsensusTimestamp());
+                        ps.setShort(2, transactionHash.getDistributionId());
+                        ps.setBytes(3, transactionHash.getHash());
+                        ps.setLong(4, transactionHash.getPayerAccountId());
+                    });
         }
+    }
+
+    private TransactionOperations getTransactionOperations() {
+        var transactionManager = new DataSourceTransactionManager(Objects.requireNonNull(jdbcTemplate.getDataSource()));
+        return new TransactionTemplate(transactionManager);
     }
 
     @Data

